File: mobile/lib/core/billing_period.dart
enum EffectiveChargeStatus { pending, overdue, paid, disregarded }

class BillingPeriod implements Comparable<BillingPeriod> {
  const BillingPeriod(this.year, this.month)
    : assert(month >= 1 && month <= 12);

  final int year;
  final int month;

  factory BillingPeriod.parse(String value) {
    final match = RegExp(r'^(\d{4})-(\d{2})$').firstMatch(value);
    if (match == null) {
      throw FormatException('Competência inválida: $value');
    }
    final year = int.parse(match.group(1)!);
    final month = int.parse(match.group(2)!);
    if (month < 1 || month > 12) {
      throw FormatException('Competência inválida: $value');
    }
    return BillingPeriod(year, month);
  }

  factory BillingPeriod.fromDate(DateTime date) =>
      BillingPeriod(date.year, date.month);

  BillingPeriod addMonths(int amount) {
    final zeroBased = year * 12 + (month - 1) + amount;
    return BillingPeriod(zeroBased ~/ 12, zeroBased % 12 + 1);
  }

  DateTime dueDate(int preferredDay) {
    final lastDay = DateTime(year, month + 1, 0).day;
    return DateTime(year, month, preferredDay.clamp(1, lastDay));
  }

  String get value =>
      '${year.toString().padLeft(4, '0')}-${month.toString().padLeft(2, '0')}';

  @override
  int compareTo(BillingPeriod other) =>
      (year * 12 + month).compareTo(other.year * 12 + other.month);

  @override
  bool operator ==(Object other) =>
      other is BillingPeriod && other.year == year && other.month == month;

  @override
  int get hashCode => Object.hash(year, month);

  @override
  String toString() => value;
}

EffectiveChargeStatus resolveChargeStatus({
  required DateTime dueDate,
  required DateTime today,
  required bool paid,
  bool disregarded = false,
}) {
  if (paid) return EffectiveChargeStatus.paid;
  if (disregarded) return EffectiveChargeStatus.disregarded;

  final due = DateTime(dueDate.year, dueDate.month, dueDate.day);
  final reference = DateTime(today.year, today.month, today.day);
  return due.isBefore(reference)
      ? EffectiveChargeStatus.overdue
      : EffectiveChargeStatus.pending;
}

String monthlyChargeDocumentId({
  required String studentId,
  required BillingPeriod period,
}) => 'mensalidade__${studentId}__${period.value}';

File: backend/src/AcademiaFight.Application/Finance/BillingPeriod.cs
using System.Globalization;

namespace AcademiaFight.Application.Finance;

public readonly record struct BillingPeriod : IComparable<BillingPeriod>
{
    public BillingPeriod(int year, int month)
    {
        if (month is < 1 or > 12)
            throw new ArgumentOutOfRangeException(nameof(month), "O mês deve estar entre 1 e 12.");

        Year = year;
        Month = month;
    }

    public int Year { get; }
    public int Month { get; }

    public string Value => $"{Year:D4}-{Month:D2}";

    public static BillingPeriod Parse(string value)
    {
        if (!DateOnly.TryParseExact(
                $"{value}-01",
                "yyyy-MM-dd",
                CultureInfo.InvariantCulture,
                DateTimeStyles.None,
                out var date))
            throw new FormatException($"Competência inválida: {value}");

        return new BillingPeriod(date.Year, date.Month);
    }

    public BillingPeriod AddMonths(int amount)
    {
        var date = new DateOnly(Year, Month, 1).AddMonths(amount);
        return new BillingPeriod(date.Year, date.Month);
    }

    public DateOnly DueDate(int preferredDay)
    {
        var lastDay = DateTime.DaysInMonth(Year, Month);
        return new DateOnly(Year, Month, Math.Clamp(preferredDay, 1, lastDay));
    }

    public int CompareTo(BillingPeriod other) =>
        (Year * 12 + Month).CompareTo(other.Year * 12 + other.Month);

    public override string ToString() => Value;
}

public enum EffectiveChargeStatus
{
    Pending,
    Overdue,
    Paid,
    Disregarded,
}

public static class ChargeStatusResolver
{
    public static EffectiveChargeStatus Resolve(
        DateOnly dueDate,
        DateOnly today,
        bool paid,
        bool disregarded = false)
    {
        if (paid) return EffectiveChargeStatus.Paid;
        if (disregarded) return EffectiveChargeStatus.Disregarded;
        return dueDate < today ? EffectiveChargeStatus.Overdue : EffectiveChargeStatus.Pending;
    }

    public static string MonthlyDocumentId(Guid studentId, BillingPeriod period) =>
        $"mensalidade__{studentId:D}__{period.Value}";
}
